File: java-spring/src/main/java/org/burgas/javaspring/mapper/Mapper.java
package org.burgas.javaspring.mapper;

import org.burgas.javaspring.dto.Request;
import org.burgas.javaspring.dto.Response;
import org.burgas.javaspring.entity.Entity;
import org.springframework.stereotype.Component;

@Component
public interface Mapper<R extends Request, E extends Entity, S extends Response, F extends Response> {

    E toEntity(R request);

    S toShortResponse(E entity);

    F toFullResponse(E entity);

    default <D> D handleData(D requestData, String message) {
        if (requestData == null) {
            throw new IllegalArgumentException(message);
        } else {
            return requestData;
        }
    }
}

File: kotlin-spring/src/main/kotlin/org/burgas/kotlinspring/mapper/Mapper.kt
package org.burgas.kotlinspring.mapper

import org.burgas.kotlinspring.entity.Model
import org.burgas.kotlinspring.entity.Request
import org.burgas.kotlinspring.entity.Response
import org.springframework.stereotype.Component

@Component
interface Mapper<in R : Request, M : Model, out S : Response, out F : Response> {

    fun toEntity(request: R): M

    fun toShortResponse(entity: M): S

    fun toFullResponse(entity: M): F
}